h/admin/services/user.py

```diff
@@ -0,0 +1,78 @@
+# -*- coding: utf-8 -*-
+
+from __future__ import unicode_literals
+
+from h import models
+from h.api.search import index
+from h.util.user import userid_from_username
+
+
+class UserRenameError(Exception):
+    pass
+
+
+class RenameUserService(object):
+    """
+    Renames a user and updates all its annotations and when necessary the NIPSA
+    status.
+
+    ``check`` should be called first
+
+    Validates the new username and updates the User. The permissions of
+    the user's annotations are updated to reflect the new username.
+
+    May raise a ValueError if the new username does not validate or
+    UserRenameError if the new username is already taken by another account.
+    """
+    def __init__(self, request):
```

Comment:
It doesn't make a lot of sense to me to pass a request object into a service. Pretty much the whole point of the service factory is to make it possible for the service to *not* require a request object, and pretty much the whole point of the service pattern is to have a layer that is *not* exposed to the internals of Pyramid.

It looks like this is ultimately because you need to construct a `BatchIndexer`. Eventually I don't think `BatchIndexer` should need a request object, but for now perhaps another way of doing this might be as follows:

```python
def rename_user_factory(context, request):
    def _reindex(ids):
        if not ids:
            return
        request.tm.commit()
        indexer = index.BatchIndexer(request.db, request.es, request)
        indexer.index(ids)
    return RenameUserService(session=request.db, auth_domain=request.auth_domain, reindex=_reindex)
```